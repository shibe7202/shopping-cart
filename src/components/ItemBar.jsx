import { useState } from "react";
import styles from './ItemBar.module.css'

function ItemBar({ data, cartTools }) {
    const [quantity, setQuantity] = useState(1)

    function decQuantity() {
        if (quantity > 1) {
            setQuantity(quantity => quantity - 1)
        }
    }

    function incQuantity() {
        if (quantity < 100) {
            setQuantity(quantity => quantity + 1)
        }
    }

    function handleAddItem() {
        cartTools.addItem({ ...data, quantity: quantity })
    }

    return (
        <div className={styles.itemBar}>
            <div className={styles.quantitySection}>
                <button className={styles.decrease} name='decrease' onClick={decQuantity}>-</button>
                <div className={styles.amount} name="quantity">{quantity}</div>
                <button className={styles.increase} name="increase" onClick={incQuantity}>+</button>
            </div>
            <div className={styles.addItemSection}>
                <button className={styles.addItemButton} onClick={handleAddItem}>Add to cart</button>
            </div>
        </div>
    )
}

export default ItemBar