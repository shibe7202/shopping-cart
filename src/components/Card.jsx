import ItemBar from "./ItemBar"
import styles from './Card.module.css'

function Card({ data, cartTools }) {

    function shortenDescription(description) {
        if (description.length < 85) {
            return description
        } else {
            let newDescription = description.slice(0, 85)
            while (!(/^[a-zA-Z]$/.test(newDescription[newDescription.length - 1]))) {
                newDescription = newDescription.slice(0, -1)
            }
            newDescription += '...'
            return newDescription
        }
    }

    return (
        <div className={styles.card}>
            <div className={styles.imgContainer}>
                <img src={data.image} alt={data.title} className={styles.img} />
            </div>
            <h2 className={styles.itemTitle}>{data.title}</h2>
            <p className={styles.itemDescription} placeholder='description'>{shortenDescription(data.description)}</p>
            <p className={styles.itemPrice}>{'$' + data.price}</p>
            <ItemBar data={data} cartTools={cartTools} />
        </div>
    )
}

export default Card