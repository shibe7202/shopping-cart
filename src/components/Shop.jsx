import useItemURL from './useItemUrl';
import Card from './Card'
import Cart from './Cart'
import NavBar from './NavBar';
import { useOutletContext } from 'react-router-dom';
import styles from './Shop.module.css'

function Shop() {
    const cartTools = useOutletContext()
    const { itemURL, error, loading } = useItemURL()

    return (
        <div className={styles.shop}>
            <NavBar cartTools={cartTools} />
            <Cart cartTools={cartTools} />
            <p className={styles.loading}>{loading ? 'Loading...' : null}</p>
            <p className={styles.error}>{error ? 'A network error has occurred' : null}</p>
            <div className={styles.cardGrid}>
                {itemURL ? itemURL.map(data =>
                    <Card data={data} cartTools={cartTools} key={data.id} />
                ) : null}

            </div>

        </div>
    )
}

export default Shop