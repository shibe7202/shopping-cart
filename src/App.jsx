import useCart from './components/useCart.jsx'
import Shop from './components/Shop.jsx'
import Home from './components/Home.jsx'
import { Outlet } from "react-router-dom";


function App() {
  const cartTools = useCart()

  return (
    <div className='mainContainer'>
      <Outlet context={cartTools} />
    </ div>
  )
}

export default App
