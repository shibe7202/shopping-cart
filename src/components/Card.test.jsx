import { vi, describe, it, expect } from "vitest";
import { render, screen, act } from "@testing-library/react";
import Shop from "./Shop";
import ItemBar from "./ItemBar.jsx";
import App from '../App.jsx'
import Card from './Card.jsx'

const data = [
    { title: 'Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops', img: '', description: 'Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday', price: 0, id: 'Item1' },
    { title: 'Mens Casual Premium Slim Fit T-Shirts', img: '', description: 'Slim-fitting style, contrast raglan long sleeve, three-button henley placket, light weight & soft fabric for breathable and comfortable wearing. And Solid stitched shirts with round neck made for durability and a great fit for casual fashion wear and diehard baseball fans. The Henley style round neckline includes a three-button placket.', price: 0, id: 'Item2' }
]

describe("Card component", () => {
    it('renders card', () => {
        const cartTools = { addItem: vi.fn() }
        const { container } = render(<Card data={data[0]} cartTools={cartTools} />)

        expect(container).toMatchSnapshot()
    })

    it('shortens long description', () => {
        const cartTools = { addItem: vi.fn() }

        act(() => render(<Card data={data[1]} cartTools={cartTools} />))

        const description = screen.getByPlaceholderText('description').textContent
        expect(description.length).toBeLessThanOrEqual(88)
        expect(description[description.length - 1]).toBe('.')
    })

})