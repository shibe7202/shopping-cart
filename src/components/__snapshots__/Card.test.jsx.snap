// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Card component > renders card 1`] = `
<div>
  <div
    class="_card_a63395"
  >
    <div
      class="_imgContainer_a63395"
    >
      <img
        alt="Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops"
        class="_img_a63395"
      />
    </div>
    <h2
      class="_itemTitle_a63395"
    >
      Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops
    </h2>
    <p
      class="_itemDescription_a63395"
      placeholder="description"
    >
      Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to...
    </p>
    <p
      class="_itemPrice_a63395"
    >
      $0
    </p>
    <div
      class="_itemBar_2ee8ab"
    >
      <div
        class="_quantitySection_2ee8ab"
      >
        <button
          class="_decrease_2ee8ab"
          name="decrease"
        >
          -
        </button>
        <div
          class="_amount_2ee8ab"
          name="quantity"
        >
          1
        </div>
        <button
          class="_increase_2ee8ab"
          name="increase"
        >
          +
        </button>
      </div>
      <div
        class="_addItemSection_2ee8ab"
      >
        <button
          class="_addItemButton_2ee8ab"
        >
          Add to cart
        </button>
      </div>
    </div>
  </div>
</div>
`;
